### ISSUE TITLE
"Prelude.undefined" error when launching simplex-chat using execSync in Node.js

### ISSUE THREAD

**Issue by andersaamodt** [simplex-chat/simplex-chat#858]
Hi! I am trying to launch simplex-chat from a menu in my terminal app, but when I call `execSync('simplex-chat')` it fails with this error:

```
simplex-chat: Prelude.undefined
CallStack (from HasCallStack):
  error, called at libraries/base/GHC/Err.hs:79:14 in base:GHC.Err
  undefined, called at platform/posix/src/System/Terminal/Platform.hsc:222:12 in terminal-0.2.0.0-B6RXs5HRbvlYycvV6QhS1:System.Terminal.Platform
```

Is this a bug? Is there maybe a workaround?

Thank you!

**Comment by epoberezkin** [comment]
Not quite sure... but it seems like you might be using the app compiled for a different platform - is it possible?

Possibly, it would help compiling it from source on your platform.

**Comment by andersaamodt** [comment]
It works fine when I launch it by typing `simplex-chat` in the terminal myself, it just doesn't work when launched from JavaScript using execSync in Node.js. Maybe there is something different about the nested execution environment that Haskell doesn't like?

**Comment by epoberezkin** [comment]
Ah - it does expect that the app can control the console, it won't work with a redirected stdout/stdin I think.

Most likely it's the keyboard input that it is missing.

What are you trying to achieve?

**Comment by andersaamodt** [comment]
I am just trying to make a menu in my Node/JavaScript command-line app where the user can select "3. Chat" from a menu and have it start Simplex Chat. Maybe I will be able to pass in args and have it automatically join a specific chatroom or set other parameters. The goal is to make it easier and more convenient for novice users to jump into a chat with zero configuration work.

**Comment by andersaamodt** [comment]
It also sets up tor and once I figure out how to connect the two, I can automate that part too.

**Comment by epoberezkin** [comment]
I think the problem is that node starts it headless. I don't know node.js well enough to say it's even possible, but the key here is to start simplex-chat in a way that node process loses any control of console and simplex-chat acquires it.

For simplex-chat to work it needs direct console key-presses, and, possibly, node.js proxies it sending input line by line? not sure, just speculating...

But looking at the error, it fails on call to `tcgetattr` in this file: https://github.com/lpeterse/haskell-terminal/blob/master/platform/posix/src/System/Terminal/Platform.hsc

**Comment by andersaamodt** [comment]
Hmm, you're right. Really what I'm trying to do is have my [menu program](https://www.npmjs.com/package/@autonomousorganization/ao-cli) exit and then auto-launch simplex-chat somehow.

I figured it out!

```
console.log('Launching Simplex Chat...')
const doNothing = () => {}
process.on('SIGINT', doNothing);
spawnSync('simplex-chat', [], { stdio: 'inherit' })
process.removeListener('SIGINT', doNothing)
```

Either spawnSync or execFileSync can be used to launch child processes from node that take over in the terminal. Ctrl-C however exits back to the terminal, but I want it to exit back into the menu system. So I captured the SIGINT signal and then remove the listener after the program returns. I guess the same Ctrl-C kills the child process but then bubbles up and is caught by Node? Anyway, it works.

Thanks for your help, you gave me the language I needed to figure this out!
